### PR TITLE
Extended Module Tracker

### PR DESCRIPTION
**Pull request by sanketpurandare**
This is an extension of [ModuleTracker](https://github.com/pytorch/pytorch/blob/main/torch/utils/module_tracker.py) with added features and bug fixes.

1. Allows installing user-defined hooks to be called in pre-fw, post-fw, pre-bw and post-bw hooks of the ``ModTracker``.
2. Adds a function ``get_known_fqn`` that retrieves the fqn of the module as tracked by the ``ModTracker``.
3. Only registers the multi-grad hooks if we are in the forward pass. This is important because, a module's pre-fw and post-fw hooks get called in the backward during AC and we do not want to register multi-grad hooks in this case. 
4. Sets the kwarg ``always_call=True`` for post-fw hooks, so that they are called post AC.

cc: @wanchaol @awgu @gnadathur 

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225 @chauhang @d4l3k

### CODE REVIEW

File: torch/distributed/_tools/mod_tracker.py [hunk 1]
```diff
@@ -0,0 +1,231 @@
...
+            post_fw_hook, self._user_post_fw_hook, "post_fw_hook"
+        )
+        self._user_pre_bw_hook = set_hook(
+            pre_bw_hook, self._user_pre_bw_hook, "pre_bw_hook"
+        )
+        self._user_post_bw_hook = set_hook(
+            post_bw_hook, self._user_post_bw_hook, "post_bw_hook"
+        )
+
+    def clear_user_hooks(self):
```
Comment (wanchaol) [inline torch/distributed/_tools/mod_tracker.py:143]: so if user want to register new hooks these are the apis that user need to call to install/clear user hooks? And I would assume these would be used by later PRs to track memory?

**Review by wanchaol — APPROVED** [review]
This feature is very interesting 🤔 

Also it looks pretty clean and easy to review, thanks for breaking the changes to smaller PR!

File: torch/distributed/_tools/mod_tracker.py (hunk 1 above)
Comment (sanketpurandare) [reply torch/distributed/_tools/mod_tracker.py:143]: Yes exactly! They can also be used to check which modules are currrently active and get their fqns.

**Comment by malfet** [comment]
Please make sure to rebase past https://github.com/pytorch/pytorch/pull/127836

**Pull request closed without merging**